File: extractors/python/devlens_extractors_python/contract.py
"""
contract Types - the JSON shape Devlens Engine expects.

Every key here is camel  case to match the typescript interface in the devlens-engine. This file is the single
source of truth for the output shape; other modules never spell keys by hand.
"""

from __future__ import annotations

from dataclasses import dataclass, field
import hashlib
from typing import Any

def code_hash(raw_code: str) -> str:
    """Return a SHA256 hash of the raw code string."""
    return hashlib.sha256(raw_code.encode("utf-8")).hexdigest()[:16]  # first 16 chars of the hash

@dataclass
class Fingerprint:
    language: str = "python"
    projectType: str = "unknown"          # "backend" once a framework is detected
    framework: str = "unknown"            # "fastapi" | "flask" | "django" | "unknown"
    router: str = "none"
    stateManagement: list = field(default_factory=list)  # always [] for Python
    dataFetching: list = field(default_factory=list)     # always [] for Python
    databases: list = field(default_factory=list)
    rawDependencies: dict = field(default_factory=dict)

    def to_dict(self) -> dict:
        return {
            "language": self.language,
            "projectType": self.projectType,
            "framework": self.framework,
            "router": self.router,
            "stateManagement": self.stateManagement,
            "dataFetching": self.dataFetching,
            "databases": self.databases,
            "rawDependencies": self.rawDependencies,
        }

@dataclass
class Stats:
    totalFiles: int = 0
    totalNodes: int = 0
    skippedFiles: int = 0

    def to_dict(self) -> dict:
        return {
            "totalFiles": self.totalFiles,
            "totalNodes": self.totalNodes,
            "skippedFiles": self.skippedFiles,
        }

@dataclass
class ExtractorError:
    file: str
    error: str

    def to_dict(self) -> dict:
        return {"file": self.file, "error": self.error}

def file_node(rel_path: str, end_line: int, node_type: str = "FILE",
              language: str = "python") -> dict:
    """FILE node — id format: file::rel/path.py (matches src/parser/index.ts)."""
    return {
        "id": f"file::{rel_path}",
        "name": rel_path.split("/")[-1],
        "type": node_type,                      # FILE | TEST
        "filePath": rel_path,
        "startLine": 1,
        "endLine": end_line,
        "parentFile": f"file::{rel_path}",      # file nodes parent themselves
        "metadata": {
            "nodeCount": 0,
            "childNodeIds": [],
            "language": language,
        },
    }

def code_node(rel_path: str, name: str, node_type: str, start_line: int, end_line: int,
              raw_code: str, metadata: dict, language: str = "python") -> dict:
    """Function/Class/Method node. ID format: rel/path.py::name."""
    return {
        "id": f"{rel_path}::{name}",
        "name": name,
        "type": node_type,
        "filePath": rel_path,
        "startLine": start_line,
        "endLine": end_line,
        "rawCode": raw_code,
        "codeHash": code_hash(raw_code),
        "parentFile": f"file::{rel_path}",
        "metadata": {"language": language, **metadata},
    }

File: extractors/java/src/devlens/extractor/Contract.java
package devlens.extractor;

import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Contract — the exact JSON shape DevLens Engine expects (mirrors
 * contract.py from the Python extractor). Every key here is camelCase to
 * match the TypeScript interfaces (CodeNode / CodeEdge / BackendRouteNode).
 * This class is the single source of truth for node/edge shapes; other
 * modules never spell keys by hand.
 */
public final class Contract {

    private Contract() {}

    /** SHA-256 of the raw code, first 16 hex chars (mirrors contract.py). */
    public static String codeHash(String rawCode) {
        try {
            MessageDigest md = MessageDigest.getInstance("SHA-256");
            byte[] digest = md.digest(rawCode.getBytes(StandardCharsets.UTF_8));
            StringBuilder sb = new StringBuilder(16);
            for (int i = 0; i < 8; i++) {
                sb.append(String.format("%02x", digest[i]));
            }
            return sb.toString();
        } catch (Exception e) {
            return "";
        }
    }

    /** FILE / TEST node — id format: file::rel/path/File.java. */
    public static Map<String, Object> fileNode(String relPath, int endLine,
                                               String nodeType, String language) {
        Map<String, Object> metadata = new LinkedHashMap<>();
        metadata.put("nodeCount", 0);
        metadata.put("childNodeIds", new ArrayList<String>());
        metadata.put("language", language);

        Map<String, Object> node = new LinkedHashMap<>();
        node.put("id", "file::" + relPath);
        node.put("name", relPath.substring(relPath.lastIndexOf('/') + 1));
        node.put("type", nodeType);
        node.put("filePath", relPath);
        node.put("startLine", 1);
        node.put("endLine", endLine);
        node.put("parentFile", "file::" + relPath);
        node.put("metadata", metadata);
        return node;
    }

    /**
     * CLASS / INTERFACE / ENUM / METHOD node — id format:
     * rel/path/File.java::TypeName or rel/path/File.java::Type.method
     */
    public static Map<String, Object> codeNode(String relPath, String name, String nodeType,
                                               int startLine, int endLine, String rawCode,
                                               Map<String, Object> metadata, String language) {
        Map<String, Object> node = new LinkedHashMap<>();
        node.put("id", relPath + "::" + name);
        node.put("name", name);
        node.put("type", nodeType);
        node.put("filePath", relPath);
        node.put("startLine", startLine);
        node.put("endLine", endLine);
        node.put("rawCode", rawCode);
        node.put("codeHash", codeHash(rawCode));
        node.put("parentFile", "file::" + relPath);

        Map<String, Object> meta = new LinkedHashMap<>();
        meta.put("language", language);
        if (metadata != null) {
            meta.putAll(metadata);
        }
        node.put("metadata", meta);
        return node;
    }

    /** THIRD_PARTY node — id format: [mvn]/prefix or [mvn]/prefix::Member. */
    public static Map<String, Object> thirdPartyNode(String id, String name,
                                                     Map<String, Object> metadata) {
        Map<String, Object> node = new LinkedHashMap<>();
        node.put("id", id);
        node.put("name", name);
        node.put("type", "THIRD_PARTY");
        node.put("filePath", id);
        node.put("startLine", 0);
        node.put("endLine", 0);
        node.put("metadata", metadata == null ? new LinkedHashMap<String, Object>() : metadata);
        return node;
    }

    /** ROUTE code node — id format: rel::METHOD /path (mirrors python common.py). */
    public static Map<String, Object> routeNode(String relPath, String httpMethod,
                                                String urlPath, boolean isDynamic,
                                                List<String> params, String framework,
                                                String handlerName) {
        Map<String, Object> metadata = new LinkedHashMap<>();
        metadata.put("urlPath", urlPath);
        metadata.put("httpMethod", httpMethod);
        metadata.put("isDynamic", isDynamic);
        metadata.put("params", params == null ? new ArrayList<String>() : params);
        metadata.put("framework", framework);
        metadata.put("handlerName", handlerName);
        metadata.put("routeKind", "backend");

        Map<String, Object> node = new LinkedHashMap<>();
        node.put("id", relPath + "::" + httpMethod + " " + urlPath);
        node.put("name", httpMethod + " " + urlPath);
        node.put("type", "ROUTE");
        node.put("filePath", relPath);
        node.put("startLine", 0);
        node.put("endLine", 0);
        node.put("parentFile", "file::" + relPath);
        node.put("metadata", metadata);
        return node;
    }

    /** BackendRouteNode dict for the `routes` array (engine BackendRouteNode). */
    public static Map<String, Object> backendRoute(String urlPath, String filePath,
                                                   String httpMethod, String framework,
                                                   boolean isDynamic, List<String> params,
                                                   String handlerName, String nodeId) {
        Map<String, Object> route = new LinkedHashMap<>();
        route.put("type", "BACKEND_ROUTE");
        route.put("urlPath", urlPath);
        route.put("filePath", filePath);
        route.put("httpMethod", httpMethod);
        route.put("framework", framework);
        route.put("isDynamic", isDynamic);
        if (params != null && !params.isEmpty()) {
            route.put("params", params);
        }
        if (handlerName != null) {
            route.put("handlerName", handlerName);
        }
        if (nodeId != null) {
            route.put("nodeId", nodeId);
        }
        return route;
    }

    /** Edge — {type, from, to, metadata?} (engine CodeEdge). */
    public static Map<String, Object> edge(String type, String from, String to,
                                           Map<String, Object> metadata) {
        Map<String, Object> e = new LinkedHashMap<>();
        e.put("type", type);
        e.put("from", from);
        e.put("to", to);
        if (metadata != null && !metadata.isEmpty()) {
            e.put("metadata", metadata);
        }
        return e;
    }

    /** Deterministic edge-key for dedupe: from|type|to. */
    public static String edgeKey(Map<String, Object> edge) {
        return edge.get("from") + "|" + edge.get("type") + "|" + edge.get("to");
    }

    public static Map<String, Object> error(String file, String message) {
        Map<String, Object> err = new LinkedHashMap<>();
        err.put("file", file);
        err.put("error", message);
        return err;
    }
}
